Update documentation (#138)
WayScience/NF1_3D_organoid_profiling_pipeline: README.md

# Neurofibromatosis Type 1 (NF1) 3D organoid image-based profiling pipeline

[![Documentation Status](https://readthedocs.org/projects/nf1-3d-organoid-profiling-pipeline/badge/?version=latest)](https://nf1-3d-organoid-profiling-pipeline.readthedocs.io/en/latest/?badge=latest)

> Patients living with Neurofibromatosis Type 1 (NF1) often develop neurofibromas

Reviewer: looks like these changes have broken the 1 sentence per line convention of markdown files. I recommend that you update.

If the precommit hook updated this, then I suggest limiting the precommit hooks privileges to skip this "fix"

Author: reverted the precommit hook in favor of another one. I fixed these edits also

> (NFs), which are complex benign tumors.

Reviewer: ```suggestion
(NFs), which are complex and benign tumors that can transform into malignant disease.
```

They are not always

### Organoid, Nuclei, Cell, and Cytoplasm Segmentations

**Organoid** | **Nuclei** | **Cell** | **Cytoplasm**
--- | --- | --- | ---
![Organoid segmentation animation](2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Organoid_mask_animation.gif) | ![Nuclei segmentation animation](2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Nuclei_mask_animation.gif) | ![Cell segmentation animation](2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Cell_mask_animation.gif) | ![Cytoplasm segmentation animation](2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Cytoplasm_mask_animation.gif)

---

## Pipeline architecture

The pipeline follows a hierarchical processing structure:

**Execution strategy:**

- SLURM-based HPC scheduling for parallel processing
- Conditional execution based on file existence
- Automatic job submission throttling (max 990 concurrent jobs)
We present a full workflow to profile 3-dimensional images of organoids.
Our end-to-end system processes raw 3D microscopy data through illumination
correction, segmentation, feature extraction, quality control, and
image-based profiling.

> ```mermaid

Reviewer: noting here the lack of `{}` wrap in mermaid diagram. The CHAMMI md file includes the wrap, which I believe you included to satisfy sphinx rendering. (I think `{}` breaks markdown rendering)

Author: removed the `{}` to avoid breaking rendering in both cases

flowchart TD
   A[Raw Microscopy Images] --> B[Stage 0: Data preprocessing]

   B --> C1[Z-stack creation]
   B --> C2["(Optional) Deconvolution"]

   C1 --> D[Deconvolved z-stack images]
   C2 --> D

   D --> E[Stage 1: Image quality control]

   E --> F1[Blur detection]
   E --> F2[Saturation dheck]

   F1 --> G[QC flags & reports]
   F2 --> G

>    G --> H[Stage 2: Image Segmentation]

Reviewer: ```suggestion
   G --> H[Stage 2: Image segmentation]
```

   (Individual feature extraction scripts)
   - Calculates specific features based on the hierarchical combination
     specified in levels 1 and 2.
   - Saves individual feature calculation outputs as a parquet file within a
     folder to be combined later.

**Feature types:**
For more details on feature types and extraction methods, refer to the
`features/` documentation.

### Feature extraction strategy

We extract hand-drawn features across multiple categories (e.g., shape,
intensity, texture) for each compartment and channel combination.
In addition, we extract deep learning-based features using the SAMMMed3D
model to capture complex morphological phenotypes that may not be described by
hand-crafted features.
SAMMed3D features are extracted as 384-dimensional embeddings per channel
per object using the CLS token output from the ViT encoder on the whole
volume.
Additionally, we take a `nucleocentric` feature extraction approach where we
extract features from a cropped volumes centered around each nucleus.
We use SAMMed3D to extract features from these nucleocentric volumes, and we
z-maximally project the nucleocentric volumes to extract 2D features using

> CHAMMI-75 features.

Reviewer: ```suggestion
CHAMMI-75.
```

each profile type:

- **Single-cell hand-drawn features** One row per nucleus/cell/cytoplasm
- **Single-cell SAMMed3D features** One row per nucleus/cell/cytoplasm
- **Organoid hand-drawn features** One row per organoid
- **Organoid SAMMed3D features** One row per organoid
- **Nucleocentric SAMMed3D features** One row per nucleus-centered volume
- **Nucleocentric CHAMMI-75 features** One row per nucleus-centered volume

Profile types:

- **Object-level** standard scalar normalized profiles (one row per object
   with all features)
- **Object-level** feature-selected profiles (one row per object with
   selected features)
- **Well-level** aggregated profiles (one row per well with aggregated
   features)
- **Consensus profiles** aggregated profiles with selected features (one row
   per treatment with selected features)

With the combination of 6 profile types and 4 profile levels, we generate a
total of 24 different profile outputs.
Each profile output is saved as a parquet file in the `data/{patient}/image_based_profiles/`.

### Image-based profiling workflow

## File and information flow diagram

1. **Feature Merging**
   - Combine all feature CSVs per well FOV

>    - Use cytotable for SQLite → Parquet conversion

Reviewer: i don't think you actually do use cytotable?

Author: Good catch. I changed this to say I am using DuckDB

    S2 --> |Feature selection| T2[Selected features]
    S3 --> |Feature selection| T3[Selected features]
    S4 --> |Feature selection| T4[Selected features]
    S5 --> |Feature selection| T5[Selected features]
    S6 --> |Feature selection| T6[Selected features]
    T1 --> U1[Aggregated profiles]
    T2 --> U2[Aggregated profiles]
    T3 --> U3[Aggregated profiles]
    T4 --> U4[Aggregated profiles]
    T5 --> U5[Aggregated profiles]
    T6 --> U6[Aggregated profiles]
    T1 --> V1[Consensus profiles]
    T2 --> V2[Consensus profiles]
    T3 --> V3[Consensus profiles]
    T4 --> V4[Consensus profiles]
    T5 --> V5[Consensus profiles]
    T6 --> V6[Consensus profiles]
```

---

Seperately from the above workflow, we also generate cross-patient consensus
profiles by merging profiles across patients and applying global feature
selection to identify a common set of features that are robust across
patients.
**Create all analysis ready output files**

```mermaid
graph LR
    A[All patient normalized profiles] --> B[Combined profiles]
    B --> C[Global feature selection]
    C --> D[Cross-patient aggregated profiles]
    D --> E[Cross-patient consensus profiles]
```

---

> **Feature Selection Parameters:**

Reviewer: ```suggestion
**Pycytominer Feature Selection Parameters:**
```

**Profile files:**

- Format: Parquet (compressed columnar storage)
- Naming: `{level}_{aggregation}.parquet`
- Example: `sc_consensus.parquet`

## Channel information

The pipeline processes four fluorescent imaging channels:
Note that while these are the channels we have used for our NF1 3D organoid
dataset, the pipeline is designed to be flexible and adaptable to other
channel configurations as needed. The channel information is stored in the
metadata and used throughout the pipeline to ensure correct processing and
feature extraction for each channel.

Name | Fluorophore | Ex(nm) | Em(nm) | Dichroic | Target | Organelle
--- | --- | --- | --- | --- | --- | ---
405 | Hoechst 33342 | 361 | 486 | 405 | DNA | Nucleus
488 | ConA Alexa Fluor 488 | 495 | 519 | 488 | ER | ER
555 | WGA Alexa Fluor 555 | 555 | 580 | 555 | Membranes | Golgi/Plasma Memb
568 | Phalloidin AF 568 | 578 | 600 | 555 | F-actin | Cytoskeleton
640 | MitoTracker Deep Red | 644 | 665 | 640 | Mitochondria | Mitochondria

**Imaging parameters:**

- Objective: 60x/1.35 NA oil immersion
- Oil RI: 1.518
- Voxel size: 0.108 μm (XY) × 1 μm (Z)
- Bit depth: 16-bit
- Dynamic range: 0-65535

## Computational specifications

> ## Hardware

Reviewer: ```suggestion
### Hardware
```

I think "computational specifications" is the H2 header, and all else follows. In other words, I recommend bumping up Hardware, local, HPC, software, environment, python utilities, system requirements, etc. by one header

### Local

- CPU: 24 cores @ 2.5 GHz
- RAM: 128 GB
- Storage: 20 TB free space
- GPU: NVIDIA GeForce 3090Ti with 24 GB VRAM for acceleration

### HPC (SLURM)

- Nodes: 100s of CPU compute nodes
- Partition: amilan (CPU), aa100 (GPU)
- QOS: normal (24h), long (7 days)
- Max concurrent jobs: 990 per user

## Software environment

### Environment Setup

> 

Reviewer: "Environment setup" instead of "Setup"

Author: implemented

We recommend using `uv`, `mamba` or `conda` to create the required
environments. We have written a `makefile` to help with conda environment
creation and management.

```bash
cd environments || exit
make --always-make
cd .. || exit
```

For `uv` users, you can also create the environments with:

```bash
source uv_setup.sh
```

### Python utilities (monorepo layout)

The utilities under `utils/src/` are now structured as installable packages.
For local development, install them in editable mode:

```bash
cd utils
pip install -e .
```

Note that the utilites should be imported into compute environments. See the
`environments` module for installing the utils. There is a Makefile in the
`environments` module that installs the environments with utils.

### System Requirements

> 

Reviewer: camel case

Author: I updated all camel/pascal cases in this document and other READMEs

    - Here we generally have 30-50 z-slices, ~1500x1500 pixel images, and 4
      channels. We rarely exceed 100 z-slices. Additionally, scaling in
      z-slices will require more compute time and RAM.
- Optional: GPU resources for segmentation and deep learning based feature
  extraction
  - We have found that a NVIDIA 3090 TI (24GB VRAM) is more than enough for
    our segmentation tasks.
  - It is important to note that part of the advantage of using 2.5D
    segmentation is that it greatly reduces the GPU VRAM requirements
    compared to full 3D segmentation - especially as z-slice count scales up.

**Storage requirements:**

- Raw images: 250-500 MB/well FOV
- Z-stacks: 250-500 MB/well FOV
- Masks: 250-500 MB/well FOV
- Features: 5-10 MB/well FOV
- Profiles: 1-5 MB/well FOV
- **Total: ~1-2 GB/well FOV**

Number of FOVs per well varies between 7-25 with typically 60 wells per
patient.
Per patient well FOVs can range from 420 to 1500 depending on the experiment
design.

**Storage estimates (per patient):**

Well FOVs | Storage (TB)
--- | ---
400 | ~0.4-0.8
500 | ~0.5-1.0
1000 | ~1.0-2.0
1500 | ~1.5-3.0

### Data Availability

> 

Reviewer: camel case

Author: see above comment

The raw and processed imaging data are not quite publicly available at this
time. We will have data available at some point on the NF Data Portal via
synapse.

## Associated repositories

For more information on the NF1 organoid profiling project, please see the
following associated repositories:

- [NF1 3D organoid profiling
  pipeline](https://github.com/WayScience/NF1_3D_organoid_profiling_pipeline)
  - This repository (code and documentation for the pipeline)
- [NF1 2D organoid profiling
  pipeline](https://github.com/WayScience/NF1_2D_organoid_profiling_pipeline)
  - Pipeline for 2D organoid profiling
- [NF1 organoid profile
  analysis](https://github.com/WayScience/NF1_organoid_profile_analysis)
  - Downstream analysis of the generated profiles

> ## Citation

Reviewer: It is great to have this section, but I highly recommend pointing folks to the citation.cff file without adding any other detail. If you add detail, then you'll have to maintain things in two places, which is bad.

Author: I implemented this and anywhere else there are citations I implented this also

---
title: Workflow: 3D organoid image-based profiling

> ---

Reviewer 2: This ends up throwing a strange error when GitHub tries to render it - I think because the frontmatter here is maybe not formatted as it expects. Consider using a hashtag-based title here.

Author: Noted and removed

> ---

Reviewer: I'm wondering why this change? Is this b/c of a sphinx expectation?

We absolutely need to address this, as it will be the first thing someone sees when they land on this project and first impressions are very important :)

Author: removed and changed title to be more descriptive

[![Documentation Status](https://readthedocs.org/projects/nf1-3d-organoid-profiling-pipeline/badge/?version=latest)](https://nf1-3d-organoid-profiling-pipeline.readthedocs.io/en/latest/?badge=latest)

> # Overview

Reviewer: ```suggestion

# Overview

```

I think this .md file needs some linting - is there a precommit hook that will handle this automatically? 

I will not make any additional formatting comments, instead relying on precommit to fix

Author: added a linter to precommit

> # Overview

Reviewer 2: H1's on a main readme should always be descriptive without much context. Consider revising this either way to make sure this content tells us what this project is about.

Author: changed!

We developed a modified 3D Cell Painting protocol to generate high-content imaging data from these organoids.
This repository contains the code and documentation for a comprehensive analysis pipeline to process and analyze these 3D organoid models of NF1 NFs.

> ## Raw channels

Reviewer: ```suggestion
## Fluorescence channels
```

Reviewer: I also recommend adding a sentence or two describing that this pipeline was developed with a specific experimental setup in mind, but that it theoretically will work in additional 3d microscopy assays

Author: added

| 405 | 488 | 555 | 640 |
|:-:|:-:|:-:|:-:|

> | <img src="2.segment_images/animations/media_for_readme/C4-2_DNA_animation.gif" alt="DNA channel animation" width="150"/> | <img src="2.segment_images/animations/media_for_readme/C4-2_ER_animation.gif" alt="ER channel animation" width="150"/> | <img src="2.segment_images/animations/media_for_readme/C4-2_AGP_animation.gif" alt="Golgi channel animation" width="150"/> | <img src="2.segment_images/animations/media_for_readme/C4-2_mitochondria_animation.gif" alt="Mito channel animation" width="150"/> |

Reviewer 2: Did you have challenges rendering these images on RTD? If so, consider using an absolute URL path to the files (sometimes external sites / docsite deployments lose track of relative paths like these). Often we can use the ["GitHub raw URL"](https://github.com/orgs/community/discussions/44370#discussioncomment-4688431) to help with this

Author: Thank you for the advice here. I will need to do this in another PR after merging the gifs as the github url does not exsist yet...

## Organoid, Nuclei, Cell, and Cytoplasm Segmentations
| Organoid | Nuclei | Cell | Cytoplasm |
|:-:|:-:|:-:|:-:|
| <img src="2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Organoid_mask_animation.gif" alt="Organoid segmentation animation" width="150"/> | <img src="2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Nuclei_mask_animation.gif" alt="Nuclei segmentation animation" width="150"/> | <img src="2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Cell_mask_animation.gif" alt="Cell segmentation animation" width="150"/> | <img src="2.segment_images/animations/media_for_readme/C4-2_NF0014_T1_C4-2_Cytoplasm_mask_animation.gif" alt="Cytoplasm segmentation animation" width="150"/> |

---

We present a full workflow to profile 3-dimensional images of organoids.

> Our end-to-end system processes raw 3D microscopy microscopy data through illumination correction, segmentation, feature extraction, and quality control.

Reviewer: ```suggestion
Our end-to-end system processes raw 3D microscopy microscopy data through illumination correction, segmentation, feature extraction, quality control, and image-based profiling.
```

> ```{mermaid}

Reviewer: something is preventing this diagram from rendering

Author: I removed the {} it renders now

> ```{mermaid}

Reviewer 2: Double checking: do you have mermaid rendering enabled within your sphinx configuration? If not, this might default to a codeblock and not a rendered artifact.

Author: Yes! Renders in the RTD

flowchart TD
   A[Raw Microscopy Images] --> B[Stage 0: Data preprocessing]

   B --> C1[Z-stack creation]
   B --> C2["(Optional) Deconvolution"]

   C1 --> D[Deconvolved z-stack images]
   C2 --> D

   D --> E[Stage 1: Image quality control]

   E --> F1[Blur detection]
   E --> F2[Saturation dheck]

   F1 --> G[QC flags & reports]
   F2 --> G

   G --> H[Stage 2: Image Segmentation]

   H --> I1[Nuclei segmentation]
   H --> I2[Cell segmentation]
   H --> I3[Organoid segmentation]

   I1 --> J[Segmentation refinement]
   I2 --> J
   I3 --> J

   J --> K[3D Segmentation masks]

   K --> L[Stage 3: Feature extraction]

   L --> M1[Area + size]
   L --> M2[Intensity]
   L --> M3[Texture]
   L --> M4[Colocalization]
   L --> M5[Neighbors]
   L --> M6[Deep learning features]

   M1 --> N[Feature matrices]
   M2 --> N
   M3 --> N
   M4 --> N
   M5 --> N
   M6 --> N

   N --> O[Stage 4: Image-based profiling]

   style B fill:#40BA40,stroke:#000,stroke-width:2px
   style E fill:#40BA40,stroke:#000,stroke-width:2px
   style H fill:#40BA40,stroke:#000,stroke-width:2px
   style L fill:#40BA40,stroke:#000,stroke-width:2px
   style O fill:#40BA40,stroke:#000,stroke-width:2px
```

> # Pipeline architecture

Reviewer: I didn't expect to see this content under this header - also, this probably should be an H2 header, reserving H1 for only the top of the doc.

Author: Only one H1 now

The pipeline follows a hierarchical processing structure:

**Execution strategy:**

- SLURM-based HPC scheduling for parallel processing
- Conditional execution based on file existence
- Automatic job submission throttling (max 990 concurrent jobs)

> # Detailed workflow stages

Reviewer: Maybe this should actually be an H3 header? Under the "architecture" H2 header?

Author: moved

## Stage 0: Data preprocessing

**Directory:** `0.preprocessing_data/`

**Purpose:** Transform raw microscopy data into standardized 3D z-stack images ready for analysis.

**Inputs:**
- Raw 2D TIFF images from microscope (5 channels × N z-slices × M wells)
- Metadata files (experiment design, plate layouts)

**Outputs:**
- 3D z-stack TIFF files organized by patient/well/FOV (optionally deconvolved)
- File structure: `data/{patient}/zstack_images/{well_fov}/{channel}.tif`

### Data preprocessing steps

1. **Patient-specific preprocessing**
   - We optimize formatting of raw images we recieve fresh from our collaborator's microscope.
   - We organize raw image files by patient ID and create patient-specific directory structures.
   - We validate file naming conventions.
2. **Updating file structure**
   - We standardize the directory hierarchy across patients.
   - We rename files to a consistent naming scheme (to account for cross-batch inconsistencies).
3. **Creating z-stacks**

>    - We combine 2D image slices into 3D z-stacks, built to accommodate different microscope formats (CQ1, Echo)

Reviewer: ```suggestion
   - We combine 2D image slices into 3D z-stacks, built to accommodate different microscope formats (e.g., CQ1, Echo)
```

   - This process maintain metadata and propagates channel information.
   - We track z-spacing between images (typically 1μm) as well as z-stack depth (between 50-100 slices [~50-100μm total])
4. **Detecting image corruption**
   - We validate TIFF file integrity removing corrupted images, incomplete stacks, and other errors introduced during image acquisition.

>    - We flag problematic datasets.

Reviewer: how is this different from "validate"? Perhaps adding some detail about what the flag means (does it trigger something? Do we handle the flag elsewhere?)

Author: added

5. **Preprocessing for deconvolution (Optional)**
   - We prepare images for Huygens deconvolution (generating parameter files and organizing batch processing structure)
6. **Post-processing for deconvolution (Optional)**
   - We import deconvolved images, verify output quality, and update file paths and metadata for downstream processing

**Key Parameters:**
- Objective: 60x/1.35 NA oil immersion
- Oil RI: 1.518

> - Voxel size: ~0.1 μm (XY) × 1 μm (Z)

Reviewer: This feels a bit out of place - perhaps list somewhere where we're detailing experimental details (near the note about how our pipeline was designed for our specific use case but in general it can work elsewhere)

Author: moved elsewhere

**Execution:**

```bash
# Example for a specific patient (NF0014)
cd 0.preprocessing_data
python scripts/1.make_zstack_and_copy_over.py --patient NF0014_T1

# Process for the CQ1 microscope
python scripts/1z.make_zstack_and_copy_over_CQ1.py --patient NF0014_T1
```

## Stage 1: Image quality control

**Directory:** `1.image_quality_control/`

**Inputs:**

> - Deconvolved z-stack images from Stage 0

Reviewer: ```suggestion
- Z-stack images from Stage 0 (deconvolution optional)
```

**Outputs:**
- QC flags file: `data/{patient}/qc_flags.csv`
- QC reports: HTML/PDF summaries with plots
- Flagged well list for exclusion from downstream analysis

**Purpose:** Assess image quality and flag problematic well FOVs before segmentation.

### Image QC steps

1. **CellProfiler QC pipeline**

>    - Extract whole-image metrics using CellProfiler.

Reviewer: ```suggestion
   - Extract whole-image QC metrics using CellProfiler.
```

>    - Compute per-slice statistics.

Reviewer: Is "statistics" different from "QC metrics"?

Author: removed

>    - Export metrics to CSV.

Reviewer: ```suggestion
   - Export QC metrics to CSV.
```

> 2. **Blur evaluation**

Reviewer: Do all downstream steps use the exported CP CSV from step 1? If so, please add this explicit detail

Author: added

   - Calculate Laplacian variance for focus detection.
   - Identify out-of-focus z-slices.
   - Set thresholds for acceptable sharpness.
3. **Saturation analysis**
   - Detect overexposed pixels per channel.
   - Calculate percentage of saturated voxels.
   - Flag wells with excessive saturation (>5%).
4. **QC report generation**
   - Create visualizations with ggplot2 (R).
   - Generate per-plate and per-patient summaries.
   - Produce pass/fail flags for each well FOV.

**Quality Metrics:**
- **Blur:** Laplacian variance, focus score

> - **Saturation:** Percentage of clipped pixels

Reviewer: ```suggestion
- **Saturation:** Percentage of pixels clipped at the maximum value
```

- **Signal-to-noise:** Mean signal / background standard deviation

> - **Illumination:** Uniformity across FOV

Reviewer: ```suggestion
- **Illumination:** Uniformity of pixel intensity across each FOV
```

**Execution:**

```bash
cd 1.image_quality_control
jupyter nbconvert --to notebook --execute notebooks/*.ipynb
```

## Stage 2: Image segmentation

**Directory:** `2.segment_images/`

> **Purpose:** Generate 3D masks for nuclei, cells, organoids, and cytoplasm compartments.

Reviewer: ```suggestion
**Purpose:** Generate 3D masks for nuclei, cells, cytoplasm, and whole organoids. We refer to each mask category as a "compartment".
```

### Image segmentation Steps

1. **Nuclei segmentation**
   - Apply Cellpose 4.0 on the DNA channel (405 nm).
2. **Organoid segmentation**
   - Use cellpose 3.x using a custom size invariant search algorithm.
3. **Cell segmentation**
   - Segment individual cells using F-actin and AGP channel (555 nm).
   - Expand from nuclear seeds.
   - Use 3D watershed for cell boundary detection.
4. **Cytoplasm derivation**
   - Subtract nuclear masks from cell masks
   - Generate cytoplasmic compartment masks
5. **Mask refinement**
   - Stitch 2D masks into 3D volumes
   - Match objects to retain the same IDs across z-slices.
   - Assign nuclei to parent cells.
   - Assign cells to parent organoids.

**Execution:**

```bash
cd 2.segment_images
sbatch grand_parent_segmentation.sh
```

## Stage 3: Feature extraction

**Directory:** `3.cellprofiling/`

> **Purpose:** Extract morphological, intensity, and texture features from segmented objects.

Reviewer: ```suggestion
**Purpose:** Extract all morphology features(e.g., shape, intensity, texture, etc.) from segmented objects.
```

Reviewer: It might also be useful to briefly describe the following:

1. How many features does the function extract?
2. How many different featurizations pipelines does stage 3 currently provide?

### Feature extraction steps

> The featurization follows a three-level hierarchical job submission structure

Reviewer: ```suggestion
To maximize parallelization and processing speed, our featurization strategy follows a three-level hierarchical job submission structure.
```

1. **Level 1: Grandparent process** (`run_featurization_grandparent.sh`)
   - Loops through all well FOVs for a patient.
   - Submits parent jobs per well FOV.
2. **Level 2: Parent process** (`run_featurization_parent.sh`)
   - Loops through feature types × compartments × channels.
   - Submits child jobs for each combination.
3. **Level 3: Child process** (Individual feature extraction scripts)
   - Calculates specific features.

>    - Saves output as parquet.

Reviewer: ```suggestion
1. **Level 1: All FOVs for a patient, per well (Grandparent process)** (`run_featurization_grandparent.sh`)
  - Submits parent jobs for each FOV (level 2).
3. **Level 2: All feature categories for each FOV (Parent process)** (`run_featurization_parent.sh`)   - Loops through all combinations of feature types × compartments × channels.
   - Submits child jobs for each feature combination (level 3).
3. **Level 3: Compute specific feature categories (Child process)** (Individual feature extraction scripts)
   - Calculates specific features based on the hierarchical combination specified in levels 1 and 2.
   - Saves individual feature calculation outputs as a parquet file within a folder to be combined later.
```

**Feature types:**
For more details on feature types and extraction methods, refer to `extraction_math` or the `features/` documentation.

## Stage 4: Image-based profiling

> **Directory:** `4.processing_image_based_profiles/`

Reviewer: why not call this simply `4.image-based-profiling`?

Author: Good point - I will refactor

> **Purpose:** Merge, normalize, and aggregate features across wells and patients ready for downstream analyses.

Reviewer: ```suggestion
**Purpose:** Merge, normalize, and aggregate features across wells and patients, preparing data for downstream analyses.
```

**Inputs:**

> - Feature CSV files from Stage 3

Reviewer: ```suggestion
- Feature parquet files from Stage 3
```

True?

> - Metadata: plate maps, treatment info, QC flags

Reviewer: ```suggestion
- Metadata: plate maps, treatment info, QC flags, etc.
```

**Outputs:**
- `data/{patient}/image_based_profiles/sc.parquet` - Single-cell profiles
- `data/{patient}/image_based_profiles/organoid.parquet` - Organoid profiles
- `data/all_patient_profiles/sc_consensus.parquet` - Cross-patient SC

> - `data/all_patient_profiles/organoid_consensus.parquet` - Cross-patient organoid

Reviewer: what does "cross patient" mean? Perhaps simply describe what you mean by consensus? Is this what you mean? If so, definition needs a slight expansion

Author: I added more information here to clarify things

- `data/all_patient_profiles/well_aggregated.parquet` - Well-level

> - `data/all_patient_profiles/patient_aggregated.parquet` - Patient-level

Reviewer: Needs expansion to describe what you mean by "well-level" and "patient-level". Perhaps the missing concept you need is "replicate"? Also, how is patient-level different from consensus?

Author: Also ripped this out and replaced it with a more in depth section

**Output Levels:**
- **Single-cell:** One row per nucleus/cell
- **Organoid:** One row per organoid (aggregated from cells)
- **Well:** One row per well FOV (aggregated from organoids)
- **Patient:** One row per patient/treatment (aggregated from wells)

> ### Image-based profiling Steps

Reviewer: ```suggestion
### Image-based profiling workflow
```

> ### Image-based profiling Steps

Reviewer: Prior to the diagram (which is a nice touch), it's probably ideal to describe how we're aggregating, how we're normalizing, and how we're feature selecting.

Reviewer: ah, I see you add this detail nicely below the diagram. I recommend moving up, prior to the diagram

Author: moved prior

## File and information flow diagram

> ```mermaid

Reviewer 2: Double checking: do you need to bracket this indicator (as you did earlier)? I wasn't sure how this might be used yet so just marking this in case helpful.

Author: I willl remove the bracket so all mermaid charts render on github from the markdown

Reviewer: Double check please as there is one that still uses brackets (the chammi doc)

Author: removed the `{}`

graph TD
    A1[CellPainting Images and Segmentations]

    A1 -->|Featurization| B[Nuclei features ]
    A1 -->|Featurization| C[Cell features ]
    A1 -->|Featurization| D[Cytoplasm features ]
    A1 -->|Featurization| E[Organoid features ]

>     A1 -->|Featurization| F[Nucleocentric features ]

Reviewer: The concept of nucleocentrism has not yet been introduced, so this is a bit jarring. I made a comment earlier about introducing the different featurization pipelines, which might be the appropriate place to introduce.

Author: introduced this as hand-drawn features in the prior section

    B --> |Merging| G[Single-cell features ]
    C --> |Merging| G[Single-cell features ]
    D --> |Merging| G[Single-cell features ]
    G --> |Annotation| G1[Single-cell features ]

    E --> |Annotation| H[Organoid features ]
    F --> |Annotation| I[Nucleocentric features ]

>     G1 --> J[Single-cell handcrafted features ]

Reviewer: i don't think we've used the term "handcrafted" in this doc before - we should use consistent nomenclature.

Author: introduced hand-drawn earlier in cellprofiling

    G1 --> K[Single-cell deep learning features ]
    H --> L[Organoid handcrafted features]
    H --> M[Organoid deep learning features]

>     I --> N[Nucleocentric Volumetric features]

Reviewer: ```suggestion
    I --> N[Nucleocentric volumetric features]
```

use consistent casing

    P5 --> |Normalization| S5[Normalized profiles]
    P6 --> |Normalization| S6[Normalized profiles]
    S1 --> |Feature selection| T1[Selected features]
    S2 --> |Feature selection| T2[Selected features]
    S3 --> |Feature selection| T3[Selected features]
    S4 --> |Feature selection| T4[Selected features]
    S5 --> |Feature selection| T5[Selected features]
    S6 --> |Feature selection| T6[Selected features]
    T1 --> U1[Aggregated profiles]
    T2 --> U2[Aggregated profiles]
    T3 --> U3[Aggregated profiles]
    T4 --> U4[Aggregated profiles]
    T5 --> U5[Aggregated profiles]
    T6 --> U6[Aggregated profiles]
    T1 --> V1[Consensus profiles]
    T2 --> V2[Consensus profiles]
    T3 --> V3[Consensus profiles]
    T4 --> V4[Consensus profiles]
    T5 --> V5[Consensus profiles]
    T6 --> V6[Consensus profiles]
```

1. **Feature Merging**
   - Combine all feature CSVs per well FOV
   - Use cytotable for SQLite → Parquet conversion
   - Create single-cell (sc) and organoid-level profiles
2. **Annotation**
   - Add treatment metadata from plate maps
   - Link drug names, targets, concentrations
   - Add patient genotype information
3. **Normalization**

>    - Z-score normalization per plate

Reviewer: is it whole plate?

Author: yes, based on dmso treated wells added int

   - Standardize features: `(x - μ) / σ`
   - Handle batch effects
4. **Feature Selection**
   - Remove low-variance features
   - Filter correlated features (correlation > 0.9)
   - Drop blocklisted features

>    - Apply frequency cutoff for categorical features

Reviewer: This section uses four different terms for the concept of removing features. I recommend being consistent - i think you mean the same thing in each case.

Author: updated

5. **Aggregation**
   - Calculate well-level statistics (mean, median, std)
   - Generate organoid-parent aggregations
   - Compute patient-level summaries

> 6. **Consensus Profiles**

Reviewer: ```suggestion
6. **Form consensus profiles**
```

   - Merge sc and organoid aggregations

>    - Create hierarchical profile structure

Reviewer: I don't know what you mean by this

Author: removed the confusiong part (second bullet)

>    - Export final consensus matrices

Reviewer: ```suggestion
   - Export final consensus profiles
```

Consistent naming

> 7. **QC Filtering**

Reviewer: ```suggestion
7. **QC filtering**
```

Reviewer: this is also a bit jarring to me that this step is so low - why are we running QC after normalization?

   - Apply image QC flags from Stage 1
   - Remove outlier objects (z-score > 3)
   - Filter low-quality wells

> 8. **Patient Combination**

Reviewer: ```suggestion
8. **Create all analysis ready output files**
```

Being more explicit about what you mean by "combination", which is confusing

   - Merge profiles across all patients

>    - Apply global feature selection

Reviewer: Are we applying feature selection twice? We should not be doing this

Author: I expanded on this - no we are not. We are doing this for the combination of all patients. I am updating the mermaid plot too

>    - Generate all-patient consensus profiles

Reviewer: We're generating consensus profiles twice?

│   │   │   ├── 3.normalized_profiles/
│   │   │   ├── 4.feature_selected_profiles/
│   │   │   └── 5.aggregated_profiles/
│   │   └── qc_flags.parquet
│   ├── NF0016_T1/
│   │   └── ... (same structure)
│   └── all_patient_profiles/
│       ├── sc_consensus.parquet
│       ├── organoid_consensus.parquet
│       ├── well_aggregated.parquet
│       └── patient_aggregated.parquet
├── models/
│   └── sam-med3d-turbo.pth
├── environments/
│   ├── GFF_preprocessing.yml
│   ├── GFF_segmentation.yml
│   └── ... (conda environments)
└── ... (code directories 0-6)
```

## File naming conventions

**Z-stack images:**
- Format: `{channel}.tif` where channel ∈ {405, 488, 555, 568, 640}
- Dimensions: (Z, Y, X)
- Data type: uint16

**Segmentation masks:**
- Format: `{compartment}_mask.tif`
- Compartments: {organoid, nuclei, cell, cytoplasm}
- Label encoding: Integer object IDs (0=background, 1-N=objects)

**Feature files:**
- Format: `{feature}_{compartment}_{channel}_{processor}_features.parquet`
- Example: `Intensity_Nuclei_405_GPU_features.parquet`

**Profile files:**
- Format: Parquet (compressed columnar storage)
- Naming: `{level}_{aggregation}.parquet`
- Example: `sc_consensus.parquet`

> # Channel information

Reviewer: i think this is duplicated from above? it also seems out of place - this should be before processing. I like referencing it (specifying how this specific setup might influence the pipeline, especially in regards to how the pipeline might be modified to accommodate different channels) here, but the info should not be duplicated.

Author: moved to live here actually. The example data takes place of this information in the top

The pipeline processes five fluorescent imaging channels:

| Name | Fluorophore | Ex(nm) | Em(nm) | Dichroic | Target | Organelle |
|------|-------------|--------|--------|----------|--------|-----------|
| 405  | Hoechst 33342        | 361    | 486    | 405      | DNA            | Nucleus           |
| 488  | ConA Alexa Fluor 488 | 495    | 519    | 488      | ER             | ER                |
| 555  | WGA Alexa Fluor 555  | 555    | 580    | 555      | Membranes      | Golgi/Plasma Memb |
| 568  | Phalloidin AF 568    | 578    | 600    | 555      | F-actin        | Cytoskeleton      |
| 640  | MitoTracker Deep Red | 644    | 665    | 640      | Mitochondria   | Mitochondria      |

**Imaging parameters:**
- Objective: 60x/1.35 NA oil immersion
- Oil RI: 1.518
- Voxel size: 0.108 μm (XY) × 1 μm (Z)
- Bit depth: 16-bit
- Dynamic range: 0-65535

> # Computational requirements

Reviewer: ```suggestion
# Computational specifications
```

I don't think we're saying this is minimum requirements, which is implied by the term "requirements"

> ## Hardware specifications

Reviewer: ```suggestion
## Hardware
```

- Profiles: 1-5 MB/well FOV
- **Total: ~1-2 GB/well FOV**

Number of FOVs per well varies between 7-25 with typically 60 wells per patient.
Per patient well FOVs can range from 420 to 1500 depending on the experiment design.

**Storage estimates (per patient):**

| Well FOVs | Storage (TB) |
|-----------|--------------|
| 400       | ~0.4-0.8     |
| 500       | ~0.5-1.0     |
| 1000      | ~1.0-2.0     |
| 1500      | ~1.5-3.0     |

### Data Availability
The raw and processed imaging data are not quite publicly available at this time.
We will have data available at some point on the NF Data Portal via synapse.

## Associated repositories
- [NF1 3D organoid profiling pipeline](https://github.com/WayScience/NF1_3D_organoid_profiling_pipeline) - This repository (code and documentation for the pipeline)
- [NF1 2D organoid profiling pipeline](https://github.com/WayScience/NF1_2D_organoid_profiling_pipeline) - Pipeline for 2D organoid profiling
- [NF1 organoid profile analysis](https://github.com/WayScience/NF1_organoid_profile_analysis) - Downstream analysis of the generated profiles

This landing page is shared with the 3D profiling pipeline repo and documentation.

> In case you are reading this on the repo landing page, you can find the documentation for the 3D profiling pipeline at: https://nf1-3d-organoid-profiling-pipeline.readthedocs.io/en/latest/?badge=latest

Reviewer: will this URL be the same always? I recommend deleting this and instead relying on sphinx/autodoc to generate and reference the link in a badge at the top

Author: removed
